# Disallow unnecessary semicolons (`no-extra-semi`)

## Rule Details

This rule extends the base [`eslint/no-extra-semi`](https://eslint.org/docs/rules/no-extra-semi) rule.
<<<<<<< HEAD

## How to use

```cjson
=======
It adds support for class properties.

## How to use

```jsonc
>>>>>>> 9fb8e1e8a00b9a1cf31203af027752dc1b545fc9
{
  // note you must disable the base rule as it can report incorrect errors
  "no-extra-semi": "off",
  "@typescript-eslint/no-extra-semi": ["error"]
}
```

<<<<<<< HEAD
=======
## Options

See [`eslint/no-extra-semi` options](https://eslint.org/docs/rules/no-extra-semi#options).

>>>>>>> 9fb8e1e8a00b9a1cf31203af027752dc1b545fc9
<sup>Taken with ❤️ [from ESLint core](https://github.com/eslint/eslint/blob/master/docs/rules/no-extra-semi.md)</sup>
