# Disallow unnecessary parentheses (`no-extra-parens`)

<<<<<<< HEAD
This rule restricts the use of parentheses to only where they are necessary.

## Rule Details

This rule extends the base [`eslint/no-extra-parens`](https://eslint.org/docs/rules/no-extra-parens) rule.
It supports all options and features of the base rule plus TS type assertions.

## How to use

```cjson
{
    // note you must disable the base rule as it can report incorrect errors
    "no-extra-parens": "off",
    "@typescript-eslint/no-extra-parens": ["error"]
=======
## Rule Details

This rule extends the base [`eslint/no-extra-parens`](https://eslint.org/docs/rules/no-extra-parens) rule.
It adds support for TypeScript type assertions.

## How to use

```jsonc
{
  // note you must disable the base rule as it can report incorrect errors
  "no-extra-parens": "off",
  "@typescript-eslint/no-extra-parens": ["error"]
>>>>>>> 9fb8e1e8a00b9a1cf31203af027752dc1b545fc9
}
```

## Options

See [`eslint/no-extra-parens` options](https://eslint.org/docs/rules/no-extra-parens#options).
<<<<<<< HEAD
=======

<sup>Taken with ❤️ [from ESLint core](https://github.com/eslint/eslint/blob/master/docs/rules/no-extra-parens.md)</sup>
>>>>>>> 9fb8e1e8a00b9a1cf31203af027752dc1b545fc9
