# Disallow unused expressions (`no-unused-expressions`)

<<<<<<< HEAD
This rule aims to eliminate unused expressions which have no effect on the state of the program.

## Rule Details

This rule extends the base [`eslint/no-unused-expressions`](https://eslint.org/docs/rules/no-unused-expressions) rule.
It supports all options and features of the base rule.
This version adds support for numerous typescript features.

## How to use

```cjson
=======
## Rule Details

This rule extends the base [`eslint/no-unused-expressions`](https://eslint.org/docs/rules/no-unused-expressions) rule.
It adds support for optional call expressions `x?.()`, and directive in module declarations.

## How to use

```jsonc
>>>>>>> 9fb8e1e8a00b9a1cf31203af027752dc1b545fc9
{
  // note you must disable the base rule as it can report incorrect errors
  "no-unused-expressions": "off",
  "@typescript-eslint/no-unused-expressions": ["error"]
}
```

## Options

See [`eslint/no-unused-expressions` options](https://eslint.org/docs/rules/no-unused-expressions#options).

<sup>Taken with ❤️ [from ESLint core](https://github.com/eslint/eslint/blob/master/docs/rules/no-unused-expressions.md)</sup>
